cmd/cosign/cli/verify/verify_blob.go

```diff
@@ -193,220 +164,100 @@ func (c *VerifyBlobCmd) Exec(ctx context.Context, blobRef string) error {
+			}
+			cert, err = loadCertFromPEM(certBytes)
+			if err != nil {
+				// check if cert is actually a public key
+				co.SigVerifier, err = sigs.LoadPublicKeyRaw(certBytes, crypto.SHA256)
 				if err != nil {
-					return fmt.Errorf("verifying certificate from bundle: %w", err)
+					return fmt.Errorf("loading verifier from bundle: %w", err)
 				}
 			} else {
-				// Verify certificate with chain
-				chain, err := loadCertChainFromFileOrURL(c.CertChain)
-				if err != nil {
-					return err
-				}
-				co.SigVerifier, err = cosign.ValidateAndUnpackCertWithChain(cert, chain, co)
-				if err != nil {
-					return fmt.Errorf("verifying certificate from bundle with chain: %w", err)
+				if c.CertChain != "" {
+					// Load certificate chain
+					chain, err = loadCertChainFromFileOrURL(c.CertChain)
+					if err != nil {
+						return err
+					}
 				}
 			}
 		}
-		if err != nil {
-			return fmt.Errorf("loading verifier from bundle: %w", err)
-		}
-		bundle = b.Bundle
-	default:
-		return fmt.Errorf("please provide a cert to verify against via --certificate or a bundle via --bundle")
-	}
-
-	// Performs all blob verification.
-	if err := verifyBlob(ctx, co, blobBytes, sig, cert, bundle); err != nil {
-		return err
+		opts = append(opts, static.WithBundle(b.Bundle))
 	}
 
-	fmt.Fprintln(os.Stderr, "Verified OK")
-	return nil
-}
-
-/* Verify Blob main entry point. This will perform the following:
-   1. Verifies the signature on the blob using the provided verifier.
-   2. Checks for transparency log entry presence:
-        a. Verifies the Rekor entry in the bundle, if provided. This works offline OR
-        b. If we don't have a Rekor entry retrieved via cert, do an online lookup (assuming
-           we are in experimental mode).
-   3. If a certificate is provided, check it's expiration.
-*/
-// TODO: Make a version of this public. This could be VerifyBlobCmd, but we need to
-// clean up the args into CheckOpts or use KeyOpts here to resolve different KeyOpts.
-func verifyBlob(ctx context.Context, co *cosign.CheckOpts,
-	blobBytes []byte, sig string, cert *x509.Certificate,
-	bundle *bundle.RekorBundle) error {
+	// Gather the cert for the signature.
+	var certPEM []byte
 	if cert != nil {
-		// This would have already be done in the main entrypoint, but do this for robustness.
-		var err error
-		co.SigVerifier, err = cosign.ValidateAndUnpackCert(cert, co)
+		certPEM, err = cryptoutils.MarshalCertificateToPEM(cert)
 		if err != nil {
-			return fmt.Errorf("validating cert: %w", err)
+			return err
+		}
+	}
+	var chainPEM []byte
+	if chain != nil {
+		// Set the last one in the co.RootCerts
+		if co.RootCerts == nil {
+			co.RootCerts = x509.NewCertPool()
+		}
+		co.RootCerts.AddCert(chain[len(chain)-1])
+		// Use the rest as the cert chain.
+		chainPEM, err = cryptoutils.MarshalCertificatesToPEM(chain)
+		if err != nil {
+			return err
 		}
 	}
+	opts = append(opts, static.WithCertChain(certPEM, chainPEM))
```

Comment:
What happens if `chain` is not set, in the case of using the default TUF chain [here](https://github.com/sigstore/cosign/blob/08277261bfdbfe8085c1a580cb3d6f59a6fea97a/cmd/cosign/cli/verify/verify_blob.go#L114)? `chainPEM` will not be set, is that an issue?

Reply:
I think the purpose was that the roots were loading in earlier, and if you have an intermediate in your chain, we'd automatically populate an intermediate pool. Had to review https://github.com/sigstore/cosign/commit/7402eb46db00a78e1aa7aec53f711a8a504d6407 where I added this

Reply:
I bet this was added originally because we didn't have the intermediate in the TUF root, so the idea was that we'd fetch the intermediate from the chain automatically (and we treated the intermediate as "untrusted", just a part of chain building)

(Item to myself, write better PR messages)

Reply:
> I think the purpose was that the roots were loading in earlier, and if you have an intermediate in your chain, we'd automatically populate an intermediate pool. Had to review https://github.com/sigstore/cosign/commit/7402eb46db00a78e1aa7aec53f711a8a504d6407 where I added this

Exactly. Before we actually reset intermediates, but in using blobs, we actually wanted to preserve the TUF roots that were populated already.